### DISCUSSION TITLE
Shapelet discovery in boolean time series

### DISCUSSION THREAD

**Discussion by Elpiro** [TDAmeritrade/stumpy#23] [Q&A]
Is the matrix profile a tool that can be used for motif discovery in boolean time series ? I am asking because the only data type accepted in stumpy.mstump() is float (no int or bool). We can workaround it by using float datatype on the 0/1 data.
But is it mathematically correct to compute the matrix profile on this kind of data ?

Usecase example : There are 4 buttons on a webpage that can be clicked in any order and as many times as you want. We could represent the actions of clicking by a multidimensional time series (4 dimensions). The patterns we're looking for are the order in which the buttons are clicked, how they are spaced in time, etc..

**Comment by seanlaw** [comment]
@Elpiro This is a great question and not something that I had considered until now. Based on what I understand from the original publications, I don’t think that this would work mathematically. Essentially, a boolean time series is the same as a time series with categorical values but only with two classes (True and False). The matrix profile needs to calculate a z-normalized Euclidean distance between two subsequences and, for a boolean time series (or categorical time series) it isn’t at all clear what the meaning of the resulting distance would be (even if the individual time series values were cast to 0.0/1.0). So, my instinct would be to advise against it but I am open to discussing it further if you like.

**Comment by seanlaw** [comment]
Perhaps, @mcyeh or @zpzim have thought about this more. I am also curious as to how matrix profile calculations would work for categorical variables (where there are more than two categories). My intuition tells me that it wouldn’t be useful for categorical cases where there isn’t a way to compare the categories but I have also been known to be wrong.

**Comment by Elpiro** [comment]
Maybe the start of an answer, the course about matrix profil here : https://www.youtube.com/watch?v=1ZHW977t070&t=1464s talks about computing the matrix profile on DNA data (at 24:24).

The way they do it is that they convert the categorical data to a real valued time series, by using this set of rules : 
~~~~
T1 = 0;
for i = 1 to length(chromosome) 
    if chromosomei = A, then Ti+1 = Ti + 2 
    if chromosomei = G, then Ti+1 = Ti + 1
    if chromosomei = C, then Ti+1 = Ti  - 1
    if chromosomei = T, then Ti+1 = Ti  - 2
end
~~~~

But I feel like it is very case-specific and wouldn't work for a set of 100+ features to create a time series from.

**Comment by seanlaw** [comment]
@Elpiro I don't think I follow the rules (and I'm a trained biochemist... whatever that is worth). It's not clear why one should add/subtract 2 or 1 and why they are mapped to specific values. Any numbers that are assigned in this way necessarily implies a quantifiable relationship between the categories.

To your point, I agree that these types of mappings are very case-specific and I wouldn't feel comfortable providing a universal approach to map categoricals to real valued time series.

**Comment by miktoki** [comment]
A comment:
I have found several use cases where using the matrix profile on arrays with very few available values, eg down to 4 integers yielded very good results, provided large window sizes(*m*), and would also like to see that it would be *possible* to do this without converting the input data to floats.

I agree with @seanlaw that it might be stretching the underlying statistical assumptions a bit far, but I personally think I would leave that decision to the user.

**Comment by seanlaw** [comment]
@miktoki There are several strong concerns here that I don't quite know how to overcome (maybe you've considered this already). When computing the matrix profile one needs to correctly compute:

1. The mean for each sliding window. It isn't clear what the mean would be for a window full of integers.
2. The standard deviation for each sliding window. It isn't clear what the stddev would be for window full of integers.
3. The Euclidean distance. It isn't clear how to compute the "distance" for integers, which requires taking a square root.

Also, the speed gain by STOMP/STUMP comes from computing the Euclidean distance and not other forms of distances (i.e., cosine, mahalanobis, manhattan). 

> it might be stretching the underlying statistical assumptions a bit far, but I personally think I would leave that decision to the user.

I don't mean to sound argumentative (so please don't take this as such) but I feel very confident in that the statistical assumptions are completely wrong (binary) if we use integers. As much as I would like to "leave the decision to the user", as a maintainer, we will be left with the repercussions of what API we choose to support. I respect your opinion so thank you for bringing this up. I want for all of us to have continued and thoughtful discussions as we prioritize new features so know that this dialogue is important to me.  

So, let's keep this feature in mind and let's collect a dozen examples (with data) to clearly demonstrate/motivate where this application would be warranted.

How does that sound?

**Comment by seanlaw** [comment]
@Elpiro @miktoki Closing this issue for now but feel free to reopen if you think that it would make sense to continue the discussion.

**Comment by MiguelGSilva** [comment]
Apologies for revisiting this discussion, but I simply wish to confirm my understanding of the issue concerning multivariate time series data comprising both numeric and categorical variables. As far as I understood, generating a multidimensional matrix profile for such cases is indeed feasible, but it needs to be through a brute force method using other distance measures, rather than employing a rapid algorithm like STUMP/STOMP which only uses Euclidean distance.

**Reply by seanlaw** [reply]
@MiguelGSilva I am very hesitant to even say that a multi-dimensional time series "comprising both numeric and categorical variables" is feasible given that, currently, after the multi-dimensional distance profile for ONE subsequence is computed for each dimension, a subset `k` of all dimensions are averaged and so there is no obvious/globally acceptable way to combine a Euclidean distance with a categorical distance. The resulting combined distance would be completely nonsensical even if you could compute some sort of categorical distance (that's a BIG "if"). 

In short, there are three problems/issues to resolve:
1. How do you compute a "categorical distance" that actually makes sense?
2. How do you compute the "categorical distance" in a matrix-profile-efficient way (i.e., allowing O(n^2) time rather than O(m*n^2) time)?
3. How do you combine a "categorical distance" with either another "categorical distance" OR a "Euclidean distance" so that it makes interpretable sense?

**Comment by NimaSarajpoor** [comment]
> Is the matrix profile a tool that can be used for motif discovery in boolean time series ? I am asking because the only data type accepted in stumpy.mstump() is float (no int or bool). We can workaround it by using float datatype on the 0/1 data. But is it mathematically correct to compute the matrix profile on this kind of data ?
> 
> Usecase example : There are 4 buttons on a webpage that can be clicked in any order and as many times as you want. We could represent the actions of clicking by a multidimensional time series (4 dimensions). The patterns we're looking for are the order in which the buttons are clicked, how they are spaced in time, etc..

I think I am late to the party :) 

I will define a problem myself and try to devise a solution (hack? trick?) 

Let's say there are different categories (or tags, activities, labels) one can record... Let's say we have five categories:
"a", "b", "c", "d" , "e"

Let's say we have a sequence `T`, containing those categories. Let's assume we get a query, Q, of length `m`, and we want to find its closest match in `T`. (We may use the approach I am about to say to compute matrix profile)

In multi-class classification, one approach is 1-vs-rest. For instance, if we have three class X, Y, Z, we will have three binary classifiers: 
* X vs YZ
* Y vs XZ
* Z vs XY 

So, basically, the first one says whether something is X or not X. We can use a similar concept here. In other words, we can do one-hot-encoding (ohe). So, in our example, the categories are `categories="abcde"`. `T` will be converted to `T_ohe`, with shape `(len(categories), len(T)`. So,  for `T[j]` containing the value `categories[i]`, `T_ohe[:, j]` is all zero except at row `i`. Then,  `stumpy.core.mass(..., normalize=False)` can be used to compute the distance profile per row, and we can just add them together. 

Example:
```
import string

import numpy as np
import stumpy


chars = string.ascii_lowercase[:5]
print(f'categories --> {chars}')

# create data with categories from chars
seed = 0
np.random.seed(seed)

IDX = np.random.choice(np.arange(len(chars)), size=100000, replace=True)
data = [chars[idx] for idx in IDX]

# one-hot-encoding 
T_data = np.zeros((len(chars), len(data)))
for i,c in enumerate(chars):
  IDX = np.flatnonzero(data == c)
  T_data[i, IDX] = 1 

# Define m, Q, and T
m = 10
Q = T_data[:, -m:]
T = T_data[:, :-m]

# compute distance profile
D = 0
for i in range(T.shape[0]):
  D = D + stumpy.core.mass(Q[i], T[i], normalize=False)

idx = np.argmin(D)
Q_NN = T[:, idx:idx + m]
```

And when we convert it back to string, we will see:

```
>>> Q
eeecbcdcec

>>> Q_1NN
eeecbcdbeb
```

As @seanlaw  pointed out:
* Does this trick make sense? :)  For each row, `i`, in our two-dimensional time series, the value 0 in its distance profile means that there is a matching pattern. The query and the discovered pattern have exact match only for the category that corresponds to that row `i`. 

* What is the proper way to compute the total distance profile? `D = D + stumpy.core.mass(Q[i], T[i], normalize=False)` 
or  `D = D + np.square(stumpy.core.mass(Q[i], T[i], normalize=False))` or ... ?

* What if we have a time series with numerical value as well? 

(Note that the underlying assumption in the proposed approach above is that the distance between any two non-identical categories is 1)
